ramips: add support for BOLT BL100

radito opened the pull request:
```
BL100 is a router made by Bolt by a provider in Indonesia
The original firmware created using Mediatek SDK and kernel 2.6.36

- Chipset: MediaTek MT7620A ver 2, eco 6
- RAM: 64 MB
- Bootloader: U Boot
- Flash: Winbond W25Q128BV (16 MB)
- Ports: 2x 10/100 Ethernet, 1x RJ-11 VoIP
- Modem: Qualcomm LTE B40 (2300 Mhz) VID:05c6 PID:9026
- Wireless: Internal MT7620A (2.4 Ghz) & Mediatek MT76x2E (5.8 Ghz)
- Switch: MediaTek MT7620A built-in 5-port 10/100M switch
- Voltage: DC 12V 1A
- Antenna Port: 2x External Antenna, 2 LTE U.FL, 2 WiFi U.FL
- Serial Port: Yes, 3.3 V TTL, Baud 57600 8N1
- Buttons: Reset and WPS
- LED: 15 Total
  - 4 blue lte, 2 red lte, 1 reset, 1 power.
  - 1 wps, 1 voip, 1 wlan2, 1 wlan5, 1 lan, 1 wan, 1 wlan.

Installation via stock firmware
1. Unlock Telnet access by downloading the backup .tar.gz file
2. Change the Telnet configuration to LAN_Telnet=1
3. Import backup configuration
4. Restart the router
5. Login to telnet with username and password = admin : db40
6. Download sysupgrade binary and mtd_write to the kernel partition
`mtd_write write openwrt-bolt_bl100-squashfs-sysupgrade.bin Kernel`
```

musashino205 commented:
Replacing the bootloader with a 3rd party one like Breed in installation instruction is not accepted on official OpenWrt.
Is it not possible to install OpenWrt while leaving the stock bootloader on your device?

radito commented:
> Replacing the bootloader with a 3rd party one like Breed in installation instruction is not accepted on official OpenWrt. Is it not possible to install OpenWrt while leaving the stock bootloader on your device?

Oh okay, sorry i just realized that is not allowed.

Need to dig deeper into the bootloader probably can, but need to validate the address more because with stock bootloader it won't boot.

Here's the original booting logs:
https://radito.github.io/posts/d52505f2-fd5b-4dd1-8286-fdc37c700cfb/

Need to work on the right address.

radito commented:
[UPDATE]
Directly writing to a kernel with mtd_write result in bad magic number, on original bootloader
Need to find a way to interrupt uBoot for tftpd


```
U-Boot 1.1.3 (Jun  8 2016 - 16:40:15)

Board:  APSoC DRAM:  64 MB
relocate_code Pointer at: 83fb8000
enable ephy clock...done. rf reg 29 = 5
SSC disabled.
spi_wait_nsec: 29 
spi device id: ef 40 18 0 0 (40180000)
find flash: W25Q128BV
raspi_read: from:30000 len:1000 
raspi_read: from:30000 len:1000 
============================================ 
LTE Router Version: 2.0.5
-------------------------------------------- 
ASIC 7620_MP (Port5<->None)
DRAM component: 512 Mbits DDR, width 16
DRAM bus: 16 bit
Total memory: 64 MBytes
Flash component: 16 MBytes NOR Flash
Date:Jun  8 2016  Time:16:40:15
============================================ 
Release Version !
raspi_read: from:30000 len:10000 
Erasing SPI Flash...
raspi_erase: offs:30000 len:10000
.
Writing to SPI Flash...
raspi_write: to:30000 len:10000 
.
done
icache: sets:512, ways:4, linesz:32 ,total:65536
dcache: sets:256, ways:4, linesz:32 ,total:32768 

 ##### The CPU freq = 580 MHZ #### 
 estimate memory size =64 Mbytes
3: System Boot system code via Flash.
## Booting image at bc050000 ...
raspi_read: from:30000 len:10000 
Erasing SPI Flash...
raspi_erase: offs:30000 len:10000
.
Writing to SPI Flash...
raspi_write: to:30000 len:10000 
.
done
raspi_read: from:50000 len:40 
Bad Magic Number,27051956 
```

radito commented:
[UPDATE] 

TFTP is enabled in uBoot but can't seem to make it load the kernel-initramfs, because of bad image file, 
i assume this is modified uboot from the isp to check image signature ?

Can't seem to figure it out.
```
Please choose the operation: 
   1: Load system code to SDRAM via TFTP. 
   2: Load system code then write to Flash via TFTP. 
   3: Boot system code via Flash (default).
   4: Entr boot command line interface.
   7: Load Boot Loader code then write to Flash via Serial. 
   9: Load Boot Loader code then write to Flash via TFTP. 

You choosed 2

 0 
raspi_read: from:40028 len:6 

   
2: System Load Linux Kernel 1 then write to Flash via TFTP. 
 Warning!! Erase Linux 1 in Flash then burn new one. Are you sure?(Y/N)
 Please Input new ones /or Ctrl-C to discard
	Input device IP (10.10.10.123) ==:10.10.10.123
	Input server IP (10.10.10.3) ==:10.10.10.3
	Input Linux Kernel filename () ==:openwrt.bin
 TIMEOUT_COUNT=10,Load address: 0x80100000
Loading: *Got ARP REPLY, set server/gtwy eth addr (xx:xx:xx:xx:xx:xx)
Got it
################Got ARP REQUEST, return our IP
Got ARP REQUEST, return our IP
#################################################
	 #################################################################
	 ###############################################################
 Same block again; ignore it 
T ##
	 #################################################################
	 #################################################################
	 #################################################################
	 #################################################################
	 #################################################################
	 #################################################################
	 #################################################################
	 #################################################################
	 ###############################################################
 Same block again; ignore it 
T ##
	 #################################################################
	 #################################################################
	 #################################################################
	 #################################################################
	 ################################################################
done
Bytes transferred = 5652071 (563e67 hex)

** Bad Image Header **
Reset !Because it's bad Image file 
```

radito commented:
[UPDATE]

After some digging and reversing the original firmware there seems some kind of magic number check that validate the kernel before it boot, need some work on that area

MAGIC_NUM="26 11 20 15"

![image](https://github.com/openwrt/openwrt/assets/11974771/88189a89-cc56-4653-97a3-84973e011d51)
![image](https://github.com/openwrt/openwrt/assets/11974771/a3768e62-6736-4418-8085-5e70e4557de1)
![image](https://github.com/openwrt/openwrt/assets/11974771/337f9eb8-a200-4b4d-90c0-31249e4154ef)

radito commented:
After hexediting magic header and calculating crc32 Openwrt boots now with the original Uboot
First 64 byte header

Magic ........CRC32.........etc
26112015 06F6FC0D 65537843 00563E27 80000000 80000000 013F7B9D 05050203 4D495053 204F7065 6E577274 204C696E 75782D35 2E31352E 31333700 00000000

```

Date:Jun  8 2016  Time:16:40:15
============================================ 
Release Version !
icache: sets:512, ways:4, linesz:32 ,total:65536
dcache: sets:256, ways:4, linesz:32 ,total:32768 

 ##### The CPU freq = 580 MHZ #### 
 estimate memory size =64 Mbytes
   
3: System Boot system code via Flash.
## Booting image at bc050000 ...
raspi_read: from:30000 len:10000 
Erasing SPI Flash...
raspi_erase: offs:30000 len:10000
.
Writing to SPI Flash...
raspi_write: to:30000 len:10000 
.
done
raspi_read: from:50000 len:40 
   Image Name:   MIPS OpenWrt Linux-5.15.137
   Image Type:   MIPS Linux Kernel Image (lzma compressed)
   Data Size:    5652007 Bytes =  5.4 MB
   Load Address: 80000000
   Entry Point:  80000000
raspi_read: from:50040 len:563e27 
   Verifying Checksum ... OK
   Uncompressing Kernel Image ... OK
No initrd
## Transferring control to Linux (at address 80000000) ...
## Giving linux memsize in MB, 64

Starting kernel ...

[    0.000000] Linux version 5.15.137 (builder@buildhost) (mipsel-openwrt-linux-musl-gcc (OpenWrt GCC 12.3.0 r23630-842932a63d) 12.3.0, GNU ld (GNU Binutils) 2.40.0) #0 Tue Nov 14 13:38:11 2023
```

john-tho commented:
> After hexediting magic header and calculating crc32 Openwrt boots now with the original Uboot First 64 byte header

You _may_ be able to do this via [`UIMAGE_MAGIC`](https://github.com/openwrt/openwrt/blob/1a6954337634474e6c42dec6023daa7af1d52432/include/image-commands.mk#L634), example: https://github.com/openwrt/openwrt/blob/1a6954337634474e6c42dec6023daa7af1d52432/target/linux/ramips/image/mt7620.mk#L1376
The OpenWrt patch for mkimage:
https://github.com/openwrt/openwrt/blob/main/tools/mkimage/patches/030-allow-to-use-different-magic.patch

radito commented:
> > After hexediting magic header and calculating crc32 Openwrt boots now with the original Uboot First 64 byte header
> 
> You _may_ be able to do this via [`UIMAGE_MAGIC`](https://github.com/openwrt/openwrt/blob/1a6954337634474e6c42dec6023daa7af1d52432/include/image-commands.mk#L634), example:
> 
> https://github.com/openwrt/openwrt/blob/1a6954337634474e6c42dec6023daa7af1d52432/target/linux/ramips/image/mt7620.mk#L1376
> 
> 
> The OpenWrt patch for mkimage:
> https://github.com/openwrt/openwrt/blob/main/tools/mkimage/patches/030-allow-to-use-different-magic.patch

Okay, thanks
Waiting on the build right now, and doing some test.

radito commented:
Update https://github.com/openwrt/openwrt/pull/14836/commits/21fcf3ce569a687d1a576aafa29c49deef488cc8

- Added UIMAGE_MAGIC for the original bootloader
- Match serial baud rate to the original bootloader
- Updated Installation Steps

john-tho commented:
You _may_ also need to use a DTS `openwrt,ih-magic` property to get a writable rootfs via a sysupgrade image:
examples: https://github.com/openwrt/openwrt/commit/ff7709d223e15cfc496abf09acae12826d9216e6
driver: https://github.com/openwrt/openwrt/blob/main/target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c

radito commented:
> You _may_ also need to use a DTS `openwrt,ih-magic` property to get a writable rootfs via a sysupgrade image: examples: [ff7709d](https://github.com/openwrt/openwrt/commit/ff7709d223e15cfc496abf09acae12826d9216e6) driver: https://github.com/openwrt/openwrt/blob/main/target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c

I was still debugging why sysupgrade image doesn't work, anyway thank you mate my mistake haha 😁

radito commented:
Update
- dts: Include uimage header
- dts: Add ih-magic value

namiltd commented:
The commit author's name in the Signed-off-by (radito) field must be your real name 'firstname lastname'

radito commented:
> The commit author's name in the Signed-off-by (radito) field must be your real name 'firstname lastname'

Okay, updated

radito commented:
@musashino205 Could i request a review for this PR ? Thanks

radito commented:
@musashino205 
> Why `initramfs-kernel`? It's not for normal use, but squashfs-sysupgrade is.

For me, it was for testing purposes so i can test that at least it can "boot" with my current build.
Should i revise it to sysupgrade for the official ?

musashino205 commented:
Yes. ```squashfs-sysupgrade``` should be tested and revised to it for normal use by users.
The changed configurations by users on ```initramfs-kernel``` won't be saved and will be disappeared when rebooting.

radito commented:
> Yes. `squashfs-sysupgrade` should be tested and revised to it for normal use by users. The changed configurations by users on `initramfs-kernel` won't be saved and will be disappeared when rebooting.

Okay, thanks
Updated for commit & desc.

musashino205 commented:
Please wrap the commit message to 74 characters or less.

https://openwrt.org/submitting-patches#submission_guidelines

> - commit description
>   - must have less than 75 characters per line

radito commented:
[> Please wrap the commit message to 74 characters or less.
> 
> https://openwrt.org/submitting-patches#submission_guidelines
> 
> > * commit description
> >   
> >   * must have less than 75 characters per line

](https://github.com/openwrt/openwrt/pull/14836#issuecomment-1999003810)

> Please wrap the commit message to 74 characters or less.
> 
> https://openwrt.org/submitting-patches#submission_guidelines
> 
> > * commit description
> >   
> >   * must have less than 75 characters per line

Okay, updated.

rmilecki commented on target/linux/ramips/dts/mt7620a_bolt_bl100.dts:
```diff
@@ -0,0 +1,258 @@
+		};
+
+		led_wlan2 {
+			label = "blue:wifi2";
```
Can you use `color` and `function` (`LED_FUNCTION_WLAN_2GHZ` I think).

radito replied:
```
Error: ../dts/mt7620a_bolt_bl100.dts:83.15-16 syntax error
FATAL ERROR: Unable to parse input tree
make[5]: *** [Makefile:235: /workspaces/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7620/image-mt7620a_bolt_bl100.dtb] Error 1
make[5]: *** Waiting for unfinished jobs....
```

The code:
```
		led_wlan2 {
			function = LED_FUNCTION_WLAN_2GHZ;
			color = <LED_COLOR_ID_BLUE>;
			gpios = <&gpio3 0 GPIO_ACTIVE_LOW>;
		};
		
		led_wlan5 {
			function = LED_FUNCTION_WLAN_5GHZ;
			color = <LED_COLOR_ID_BLUE>;
			gpios = <&gpio1 5 GPIO_ACTIVE_LOW>;
		};
		
```

@rmilecki 
I'm getting this error when trying to build it, invalid input tree ?

radito replied:
Is it because the patch is new https://github.com/openwrt/openwrt/commit/081cd769fc3e1de7709d7473b2560aaa549e2b05

radito replied:
Updated

rmilecki commented on target/linux/ramips/dts/mt7620a_bolt_bl100.dts:
```diff
@@ -0,0 +1,258 @@
+		};
+		
+		led_wlan5 {
+			label = "blue:wifi5";
```
Can you use `color` and `function` (`LED_FUNCTION_WLAN_5GHZ` I think).

radito replied:
Updated

hauke commented on target/linux/ramips/dts/mt7620a_bolt_bl100.dts:
```diff
@@ -0,0 +1,260 @@
+		led_wlan5 {
+			function = LED_FUNCTION_WLAN_5GHZ;
+			color = <LED_COLOR_ID_BLUE>;
+			gpios = <&gpio1 5 GPIO_ACTIVE_LOW>;
```
Does this change work instead of adding the wifi trigger in target/linux/ramips/mt7620/base-files/etc/board.d/01_leds ?

```suggestion
			gpios = <&gpio1 5 GPIO_ACTIVE_LOW>;
			linux,default-trigger = "phy0tpt";
```

radito replied:
Updated

hauke commented on target/linux/ramips/dts/mt7620a_bolt_bl100.dts:
```diff
@@ -0,0 +1,260 @@
+		led_wlan2 {
+			function = LED_FUNCTION_WLAN_2GHZ;
+			color = <LED_COLOR_ID_BLUE>;
+			gpios = <&gpio3 0 GPIO_ACTIVE_LOW>;
```
Does this change work instead of adding the wifi trigger in target/linux/ramips/mt7620/base-files/etc/board.d/01_leds ?

```suggestion
			gpios = <&gpio3 0 GPIO_ACTIVE_LOW>;
			linux,default-trigger = "phy1tpt";
```

radito replied:
Yes, i can confirm it works.
Tested with latest main branch.

radito replied:
Updated

radito committed:
ramips: add support for BOLT BL100

BL100 is a router made by Bolt by a provider in Indonesia
The original firmware created using Mediatek SDK and kernel 2.6.36

- Chipset: MediaTek MT7620A ver 2, eco 6
- RAM: 64 MB
- Bootloader: U Boot
- Flash: Winbond W25Q128BV (16 MB)
- Ports: 2x 10/100 Ethernet, 1x RJ-11 VoIP
- Modem: Qualcomm LTE B40 (2300 Mhz) VID:05c6 PID:9026
- Wireless: Internal MT7620A (2.4 Ghz) & Mediatek MT76x2E (5.8 Ghz)
- Switch: MediaTek MT7620A built-in 5-port 10/100M switch
- Voltage: DC 12V 1A
- Antenna Port: 2x External Antenna, 2 LTE U.FL, 2 WiFi U.FL
- Serial Port: Yes, 3.3 V TTL, Baud 57600 8N1
- Buttons: Reset and WPS
- LED: 15 Total
  - 4 blue lte, 2 red lte, 1 reset, 1 power.
  - 1 wps, 1 voip, 1 wlan2, 1 wlan5, 1 lan, 1 wan, 1 wlan.

Installation via stock firmware
1. Unlock Telnet access by downloading the backup .tar.gz file
2. Change the Telnet configuration to LAN_Telnet=1
3. Import backup configuration
4. Restart the router
5. Login to telnet with username and password = admin : db40
6. Download sysupgrade binary and mtd_write to the kernel partition
`mtd_write write openwrt-bolt_bl100-squashfs-sysupgrade.bin Kernel`

Signed-off-by: Radito Wahyu <arditogits@gmail.com>

radito commented:
Just wondering if this will make it to next realease @hauke ?, really looking forward for this device 

Because this is not included in [23.05.4](https://downloads.openwrt.org/releases/23.05.4/targets/)

rany2 commented:
@radito stable releases rarely add support for new devices, only bug fixes.
